### PR TITLE
Semi-Additive Measure

### PR DESCRIPTION
**Pull request by WilliamDee**
This PR adds the ability to specific dimensions that are non additive in a measure, otherwise known as a semi-additive measure. This is used when we can aggregate a measure on certain dimensions, but not on other dimensions. A way to denote these measures would be as followed in the config:
```yaml
data_source:
   ...
   measures:
      ...
      non_additive_dimensions:
         - name: <dimension_name> # Currently only TimeDimension are allowed to be specified as a non_additive dimension
           window_choice: # one of ["min"/"max"] 
           window_groupings: # <optional list of identifiers to join on>
```

The underlying sql rendering here is filtering the data source down to singular values via an aggregate function for the non-additive dimension then joining it back to the data set.

For example if we look at the following config,
```yaml
data_source: 
  measures:
    - name: average_account_balance_specified
      agg: avg
      non_additive_dimensions:
        - name: statement_date
          window_choice: min
``` 
this would be rendered out as
```sql
SELECT AVG(average_account_balance_specified), ...
FROM
   <data_set_1> ds1
INNER JOIN
   SELECT MIN(statement_date) AS statement_date
   FROM 
   <data_set_1> ds2
ON
  ds1.statement_date = ds2.statement_date
...
```
In addition, one can specify additional group by identifiers for that right join statement.
```yaml
data_source: 
  measures:
    - name: average_account_balance_specified
      agg: avg
      non_additive_dimensions:
        - name: statement_date
           window_choice: min
           window_grouping:
              - customer_id
``` 
we would get this,
```sql
SELECT AVG(average_account_balance_specified), ...
FROM
   <data_set_1> ds1
INNER JOIN
   SELECT MIN(statement_date) AS statement_date, customer_id
   FROM 
   <data_set_1> ds2
   GROUP BY customer_id
ON
  ds1.statement_date = ds2.statement_date AND ds1.customer_id = ds2.customer_id
...
```

### CODE REVIEW

File: metricflow/dataflow/builder/source_node.py
Comment (tlento) [inline metricflow/dataflow/builder/source_node.py:52]: There's a lot of cross-checking gymnastics in this PR. If we make the non-additive attributes part of the MeasureSpec instead of keeping them separated in DataSourceSemantics, does that simplify the other things?

I think it should work to do that although it is a larger refactor underneath this to update the various direct calls to the MeasureSpec initializer.

File: metricflow/dataflow/builder/source_node.py
Comment (tlento) [inline metricflow/dataflow/builder/source_node.py:80]: Ok this seems right to me but I'll defer to @plypaul on whatever he might turn up here.

From a code structure perspective, I think it'll be easier to understand the block contained in this inner for loop if a lot of this was moved into a helper function to initialize and return a MetricTimeDimensionTransformNode.

**Review by tlento — COMMENTED** [review]
Looks promising!

File: metricflow/specs.py
Comment (WilliamDee) [inline metricflow/specs.py:319]: Is there another way around this?

File: metricflow/dataflow/dataflow_plan.py
Comment (plypaul) [inline metricflow/dataflow/dataflow_plan.py:742]: Instead of hiding the description, what about a different description that better describes the operation You were looking to filter out measures that have different additive properties?

File: metricflow/model/objects/elements/measure.py
Comment (plypaul) [inline metricflow/model/objects/elements/measure.py:56]: For a follow up, we'll want to have a rule to validate these items.

File: metricflow/model/semantics/semantic_containers.py
Comment (plypaul) [inline metricflow/model/semantics/semantic_containers.py:154]: This nesting is getting a little complicated - maybe introduce a helper object?

File: metricflow/dataflow/builder/source_node.py
Comment (plypaul) [inline metricflow/dataflow/builder/source_node.py:53]: Mind updating the comment to describe how semi-additive measures are handled?

**Review by plypaul — COMMENTED** [review]
Generally looks good, but it would be great to see some test cases that demonstrate how the node gets rendered into a SQL query.

File: metricflow/specs.py
Comment (WilliamDee) [reply metricflow/specs.py:319]: Yup, it errors out with circular import if I put this to the top where the imports should be. Putting it where I put it in this PR works, but I didn't think it was a good idea to put it there unless it's actually okay.

File: metricflow/dataflow/dataflow_plan.py
Comment (WilliamDee) [reply metricflow/dataflow/dataflow_plan.py:742]: replaced with `replace_description` where we would pass a string to replace the current description! For example, here when we filter only semi additive measures it would say,
```
<FilterElementsNode>
                <!-- description = Filter all Semi-additive Measures -->
                <!-- node_id = pfe_6 -->
```

File: metricflow/dataflow/builder/costing.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/costing.py:160]: followed that, but did this 
```python
parent_costs = [x.accept(self) for x in node.parent_nodes]
 # Add number of joins to the cost.
node_cost = DefaultCost(num_joins=1)
return DefaultCost.sum(parent_costs + [node_cost])
```
to accomodate for the join. Is this correct?

File: metricflow/model/objects/elements/measure.py
Comment (WilliamDee) [reply metricflow/model/objects/elements/measure.py:67]: agreed to use sha1 through discussions!

File: metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml
Comment (plypaul) [inline metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml:17]: You can remove the `SELECT *` and just do `{{ source_schema }}.fct_accounts`. Same for the other cases where you do `SELECT *`.

File: metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml
Comment (plypaul) [inline metricflow/test/integration/test_cases/itest_semi_additive_measure.yaml:32]: Guessing this might be from generated SQL - parenthesis are in there because we haven't put in operator precedence logic, but since people will be reading these, you can remove the parenthesis.

**Review by plypaul — COMMENTED** [review]
Almost ready to approve, but seeing if you saw my comment about the nested `Dict` structure?

File: metricflow/dataflow/builder/dataflow_plan_builder.py
Comment (plypaul) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:514]: In retrospect, `build_aggregated_measures_for_single_data_source` is better named to something like `build_aggregated_measures_from_measure_source_node` now. What do you think?

**Review by plypaul — APPROVED** [review]
Great work on this feature! Left some nits.

File: metricflow/model/semantics/element_group.py
Comment (WilliamDee) [reply metricflow/model/semantics/element_group.py:10]: no reason, going to do `Element pair` instead

File: metricflow/model/objects/elements/measure.py
Comment (WilliamDee) [reply metricflow/model/objects/elements/measure.py:46]: changed

File: metricflow/dataflow/builder/source_node.py
Comment (WilliamDee) [reply metricflow/dataflow/builder/source_node.py:39]: Made it similar to the original description in `FilterElementsNode` ie., `Pass only Semi-Additive Measures`

File: metricflow/test/fixtures/model_yamls/simple_model/metrics.yaml
Comment (WilliamDee) [reply metricflow/test/fixtures/model_yamls/simple_model/metrics.yaml:338]: Changed all addresses in this file to support@transformdata.io

File: metricflow/dataflow/dataflow_plan.py
Comment (WilliamDee) [reply metricflow/dataflow/dataflow_plan.py:421]: Changed to this, please review as I'm not too familiar with the data lingo here.

```
A node that performs a row filter by aggregating a given non-additive dimension.

    This is designed to filter a dataset down to singular non-additive time dimension values by aggregating
    the time dimension with MAX/MIN then joining back to the original dataset on that aggregated value.
    Additionally, an optional sequence of identifiers can be passed in to group by and join on during the filtering.

    For example, if we have a data set that includes "account_balances,user,timestamp", we can build a
    "latest_account_balance_by_user" data set using this node by filtering the data set such that for each user,
    we get the MAX(date) and return the account_balance corresponding to that date.
```

Comment (plypaul) [reply metricflow/dataflow/dataflow_plan.py]: That looks good, but how about a small table with the columns account_balances, user, timestamp shown before and after the?

Comment (WilliamDee) [reply metricflow/dataflow/dataflow_plan.py]: Added!

**Pull request merged**
